### PR TITLE
Making py chip-device-ctrl discoverable over DNS-SD as a commissioner

### PR DESCRIPTION
**Pull request by sharadb-amazon**
#### Problem
For a Matter Test event, we need to run the python chip-device-ctrl tool on a raspi and test the tv-casting-app against it. For this the tool needs to be discoverable over DNS-SD as a commissioner (It already accepts UDC requests and has the connect command to do onnetwork commissioning)

#### Change overview
With this change, the python chip-device-ctrl tool will start advertising itself as a commissioner on start-up behind a feature flag: CHIP_DEVICE_CONFIG_ENABLE_COMMISSIONER_DISCOVERY.
TBD: Code to set the flag CHIP_DEVICE_CONFIG_ENABLE_COMMISSIONER_DISCOVERY to 1 for just the python chip-device-ctrl tool will follow separately.

#### Testing
Tested with the CHIP_DEVICE_CONFIG_ENABLE_COMMISSIONER_DISCOVERY set to 1. Used the minimal-mdns client to ensure the python chip-device-ctrl tool is discoverable as a commissioner device

### CODE REVIEW

File: src/app/server/Mdns.cpp [hunk 1]
```diff
@@ -246,7 +241,18 @@ CHIP_ERROR Advertise(bool commissionableNode, CommissioningMode mode)
             advertiseParameters.SetPairingInstr(chip::Optional<const char *>::Value(pairingInst));
         }
     }
+    return CHIP_NO_ERROR;
+}
 
+/// Overloaded utility method for commissioner and commissionable advertisement
+/// This method is used for both commissioner discovery and commissionable node discovery since
+/// they share many fields.
+///   commissionableNode = true : advertise commissionable node
```
Comment (andy31415) [inline src/app/server/Mdns.cpp:250]: true/false seems to be hard to follow: I read `Advertise(true)` as "yes, advertise" and `Advertise(false)` as 'no, do not advertise.

Please use an enum like `kCommisionableNode` and `kCommisioner` or similar (i.e. signature should look like `Advertise(AdvertiseNodeType nodeType, CommisioningMode mode)`

File: src/app/server/Mdns.cpp [hunk 2]
```diff
@@ -246,7 +241,18 @@ CHIP_ERROR Advertise(bool commissionableNode, CommissioningMode mode)
...
+///   commissionableNode = false : advertise commissioner
+CHIP_ERROR Advertise(bool commissionableNode, CommissioningMode mode)
+{
+    auto advertiseParameters = chip::Mdns::CommissionAdvertisingParameters();
```
Comment (andy31415) [inline src/app/server/Mdns.cpp:254]: why the auto? Why not:

```cpp
chip::Mdns::CommissionAdvertisingParameters advertiseParameters;
```

otherwise this seems to make it seem as you are calling a method with some unknown parameters.

The method seems to create a variable on the stack and then call build on it, never using it again. Why is that? if the argument is never used, why do we need to pass in a reference?

File: src/app/server/Mdns.h [hunk 3]
```diff
@@ -52,6 +53,9 @@ CHIP_ERROR AdvertiseCommissioner();
 /// Set MDNS commissionable node advertisement
 CHIP_ERROR AdvertiseCommissionableNode(CommissioningMode mode);
 
+CHIP_ERROR BuildCommissionAdvertisingParameters(bool commissionableNode, CommissioningMode mode,
```
Comment (andy31415) [inline src/app/server/Mdns.h:56]: Place doxygen comments in the header not in the cpp file

File: src/app/server/Mdns.cpp [hunk 4]
```diff
@@ -137,19 +136,15 @@ CHIP_ERROR AdvertiseOperational()
     return CHIP_NO_ERROR;
 }
 
-/// Overloaded utility method for commissioner and commissionable advertisement
-/// This method is used for both commissioner discovery and commissionable node discovery since
-/// they share many fields.
 ///   commissionableNode = true : advertise commissionable node
 ///   commissionableNode = false : advertise commissioner
-CHIP_ERROR Advertise(bool commissionableNode, CommissioningMode mode)
+CHIP_ERROR BuildCommissionAdvertisingParameters(bool commissionableNode, CommissioningMode mode,
```
Comment (andy31415) [inline src/app/server/Mdns.cpp:141]: document what this does (not here, in the header).

**Comment by msandstedt** [comment]
I know we're trying to get things done for test events.  But it would be really nice if we addressed some other closely related, underlying issues:

https://github.com/project-chip/connectedhomeip/issues/6141
https://github.com/project-chip/connectedhomeip/issues/7911
https://github.com/project-chip/connectedhomeip/issues/8438
https://github.com/project-chip/connectedhomeip/issues/8722

If we're touching this code, can't some of the above issues get some attention too?

File: src/controller/python/ChipDeviceController-ScriptBinding.cpp [hunk 5]
```diff
@@ -210,6 +211,13 @@ ChipError::StorageType pychip_DeviceController_NewDeviceController(chip::Control
     initParams.controllerNOC                  = nocSpan;
 
     (*outDevCtrl)->SetUdpListenPort(CHIP_PORT + 1);
+#if CHIP_DEVICE_CONFIG_ENABLE_COMMISSIONER_DISCOVERY // make this commissioner discoverable
+    auto advertiseParameters = chip::Mdns::CommissionAdvertisingParameters();
+    chip::app::Mdns::BuildCommissionAdvertisingParameters(false, chip::app::Mdns::CommissioningMode::kDisabled,
+                                                          advertiseParameters);
+    (*outDevCtrl)->SetCommissionAdvertisingParameters(&advertiseParameters);
```
Comment (Damian-Nordic) [inline src/controller/python/ChipDeviceController-ScriptBinding.cpp:218]: If this is meant to be used only during the initialization, move it to `CommissionerInitParams`. Otherwise, it's unsafe to store a pointer to a local variable.

File: src/controller/python/ChipDeviceController-ScriptBinding.cpp [hunk 6]
```diff
@@ -210,6 +211,13 @@ ChipError::StorageType pychip_DeviceController_NewDeviceController(chip::Control
     initParams.controllerNOC                  = nocSpan;
 
     (*outDevCtrl)->SetUdpListenPort(CHIP_PORT + 1);
+#if CHIP_DEVICE_CONFIG_ENABLE_COMMISSIONER_DISCOVERY // make this commissioner discoverable
+    auto advertiseParameters = chip::Mdns::CommissionAdvertisingParameters();
+    chip::app::Mdns::BuildCommissionAdvertisingParameters(false, chip::app::Mdns::CommissioningMode::kDisabled,
```
Comment (Damian-Nordic) [inline src/controller/python/ChipDeviceController-ScriptBinding.cpp:216]: You don't seem to pass any non-default options which `AdvertiseCommissioner` would not set. Why is this function being exposed then?

**Review by Damian-Nordic — CHANGES_REQUESTED** [review]

**Comment by sharadb-amazon** [comment]
https://github.com/project-chip/connectedhomeip/pull/9348 made some significant simplifications to how a commissioner can be advertised. So, the old diff in this PR is moot. Now this PR is just a 4 line change. See https://github.com/project-chip/connectedhomeip/pull/9603/commits/2af778ceed8515c53092c903d5a41fe100c17e57

**Review by Damian-Nordic — APPROVED** [review]

**Review by msandstedt — APPROVED** [review]

**Review by chrisdecenzo — APPROVED** [review]

File: src/controller/python/ChipDeviceController-ScriptBinding.cpp [hunk 7]
```diff
@@ -214,6 +215,9 @@ ChipError::StorageType pychip_DeviceController_NewDeviceController(chip::Control
     (*outDevCtrl)->SetUdpListenPort(0);
     err = (*outDevCtrl)->Init(initParams);
     VerifyOrReturnError(err == CHIP_NO_ERROR, err.AsInteger());
+#if CHIP_DEVICE_CONFIG_ENABLE_COMMISSIONER_DISCOVERY
```
Comment (andy31415) [inline src/controller/python/ChipDeviceController-ScriptBinding.cpp:218]: Will this require separate compile options to have the python build in one mode vs the other? How does this work? 

I have a hard time matching PR description with change contents.

Comment (cecille) [reply src/controller/python/ChipDeviceController-ScriptBinding.cpp:218]: Because we have these flags in the codebase anyway, we're already de-facto compiling with different flags. They're plumbed through gn normally, but we can compile with ex. different mdns impls etc. But I don't actually think this needs to be Ifdef'd - it seems like you can call StartServer and it will start advertising as a commissioner if that flag is on.

However, I agree with Andrei - it's unclear how you would compile the python controller to enable or disable this functionality.

**Comment by cecille** [comment]
> I know we're trying to get things done for test events. But it would be really nice if we addressed some other closely related, underlying issues:
> 
> #6141
> #7911
> #8438
> #8722
> 
> If we're touching this code, can't some of the above issues get some attention too?

Some of those issues are duplicates that were never actually assigned. Ex. multiple operational adverts are supported and have been tested. Please see TestAdvertiser.cpp and TestPlatform.cpp.

File: src/controller/python/ChipDeviceController-ScriptBinding.cpp (hunk 7 above)
Comment (woody-apple) [reply src/controller/python/ChipDeviceController-ScriptBinding.cpp:218]: Why is the python controller required here?

Comment (woody-apple) [reply src/controller/python/ChipDeviceController-ScriptBinding.cpp:218]: Filed this:
https://github.com/project-chip/connectedhomeip/issues/9743
